src/stochastic_search_iterator.jl

```diff
@@ -0,0 +1,174 @@
+
+
+    @info "Start: $(rulenode2expr(current_program, grammar)), subexpr: $(rulenode2expr(subprogram, grammar)), cost: $current_cost
+            temp $new_temperature"
+
+    # propose new programs to consider. They are programs to put in the place of the node
+    possible_replacements = iter.propose(current_program, neighbourhood_node_location, grammar, iter.max_depth, dict)
+    
+    # the next program in the iteration
+    next_program = deepcopy(current_program)
+    possible_program = current_program
+    best_replacement = nothing
+    # @info "Possible replacements size: $(length(possible_replacements))"
+    for possible_replacement in possible_replacements
+        # replace node at node_location with new_random 
+        if neighbourhood_node_location.i == 0
+            possible_program = possible_replacement
+        else
+            # update current_program with the subprogram generated
+            # this line mutates also the current_program. That is why we deepcopy at 115
+            neighbourhood_node_location.parent.children[neighbourhood_node_location.i] = possible_replacement
+        end
+        program_cost = calculate_cost(possible_program, iter.cost_function, examples, grammar)
+        # TODO: check whether it should be previous or new temperature
+        if iter.accept(current_cost, program_cost, new_temperature) 
+            next_program = deepcopy(possible_program)
+            current_cost = program_cost
+            best_replacement = deepcopy(possible_replacement)
+        end
+    end
+
+    # if best_replacement !== nothing
+    #     @info "Best replace: $(rulenode2expr(best_replacement, grammar)) => Cost : $current_cost"
+    # else
+    #     @info "Can't find better"
+    # end
+    # @info "================"
+
+    if current_cost < current_state.best_program_cost
+        next_state = IteratorState(
+            current_program=next_program,
+            current_temperature=new_temperature,
+            best_program=next_program,
+            best_program_cost=current_cost)
+    else
+        next_state = IteratorState(
+            current_program=next_program,
+            current_temperature=new_temperature,
+            best_program=current_state.best_program,
+            best_program_cost=current_state.best_program_cost)
+    end
+
+    return (next_program, next_state)
+end
+
+"""
+Returns the cost of the `program` using the examples and the `cost_function`. It first convert the program to an expression and
+evaluates it on all the examples.
+"""
+function calculate_cost(program::RuleNode, cost_function::Function, examples::AbstractVector{Example}, grammar::Grammar)
+    results = Tuple{Int64,Int64}[]
+    expression = rulenode2expr(program, grammar)
+    symbol_table = SymbolTable(grammar)
+    for example ∈ filter(e -> e isa IOExample, examples)
+        outcome = HerbEvaluation.test_with_input(symbol_table, expression, example.in)
```

Comment:
This only uses the default Julia evaluator. It would be nice if the algorithm could take any evaluation function, just like line 19 in the new search procedure: https://github.com/Herb-AI/HerbSearch.jl/blob/b220b90fa87ef7dbb547a0377c55b94876a8ddc8/src/search_procedure.jl#L15-L24

Reply:
We saw that in Dev branch there are functions `search` and `search_best` that take the evaluator. We could make use of those to include evaluation function in our algorithm after Dev is merged.

Reply:
I agree, I would also do that.